src/metrics.rs

```diff
@@ -83,57 +66,70 @@ impl Metrics {
     /// drop(connection_close);
     /// ```
     pub fn increment_defer<'a, M1, M2>(
-        self: &Arc<Metrics>,
+        &'a self,
         event: &'a M1,
-    ) -> Deferred<'a, impl FnOnce(Arc<Metrics>)>
+    ) -> Deferred<'a, impl FnOnce(&'a Self), Self>
     where
         M1: Clone + 'a,
         M2: From<&'a M1> + 'a,
         Metrics: IncrementRecorder<M1> + IncrementRecorder<M2>,
     {
         self.increment(event);
         let m2: M2 = event.into();
-        self.defer(move |metrics| {
+        self.defer_record(move |metrics| {
             metrics.increment(&m2);
         })
     }
 }
 
-pub struct Deferred<'a, F>
+pub struct Deferred<'a, F, T>
 where
-    F: FnOnce(Arc<Metrics>),
+    F: FnOnce(&'a T),
+    T: ?Sized,
 {
-    metrics: Arc<Metrics>,
-    record_fn: Option<F>,
-    _lifetime: PhantomData<&'a F>,
+    param: &'a T,
+    deferred_fn: Option<F>,
 }
 
-impl<'a, F> Deferred<'a, F>
+impl<'a, F, T> Deferred<'a, F, T>
 where
-    F: FnOnce(Arc<Metrics>),
+    F: FnOnce(&'a T),
+    T: ?Sized,
 {
-    pub fn new(metrics: Arc<Metrics>, record_fn: F) -> Self {
+    pub fn new(param: &'a T, deferred_fn: F) -> Self {
         Self {
-            metrics,
-            record_fn: Some(record_fn),
-            _lifetime: PhantomData::default(),
+            param,
+            deferred_fn: Some(deferred_fn),
         }
     }
 }
 
-impl<'a, F> Drop for Deferred<'a, F>
+impl<'a, F, T> Drop for Deferred<'a, F, T>
 where
-    F: FnOnce(Arc<Metrics>),
+    F: FnOnce(&'a T),
+    T: ?Sized,
 {
     fn drop(&mut self) {
-        if let Some(record_fn) = mem::take(&mut self.record_fn) {
-            (record_fn)(self.metrics.clone());
+        if let Some(deferred_fn) = mem::take(&mut self.deferred_fn) {
+            (deferred_fn)(self.param);
         } else {
-            error!("defer record failed, event is gone");
+            error!("defer deferred record failed, event is gone");
         }
     }
 }
 
+pub trait DeferRecorder {
```

Comment:
meant for another PR?

Reply:
I had to extract this from `Metrics` because I separated out the DNS metrics. Now both structures implement this.